server/src/main/java/org/elasticsearch/common/util/CachedSupplier.java

```diff
@@ -38,14 +40,18 @@ public T get() {
         if (supplier == null) {
             return result;
         }
-        initResult();
-        return result;
+        return initResult();
     }
 
-    private synchronized void initResult() {
-        if (supplier != null) {
-            result = supplier.get();
+    private synchronized T initResult() {
+        var s = supplier;
+        if (s != null) {
+            T res = s.get();
+            result = res;
```

Comment:
Nit: should we add a comment here to not reorder these? To preserve the same guarantees, we want the volatile write to be where it is, and no one should be tempted to do `supplier = null; return result = s.get()` or something similar.

Reply:
I added the explanation for this to the field declarations. I would tend to not document this inline. It seems like this shouldn't need more than one comment. No one should "just" reorder variable access in threadsafe code. Maybe we should look into a scheme for marking these spots consistently instead? (annotation on the method on inline or so ... even if that's overlooked it at least provides a short+sweet+standard hint?) What's the core-infra opinion of that? :)